libkatana_python_native/include/katana/python/NumbaSupport.h

```diff
@@ -0,0 +1,161 @@
+  /// function pointer. Unlike methods, there is no need for a data pointer.
+  template <typename Return, typename... Extra>
+  void def_func(
+      pybind11::module_& m, const char* name, Return (*f)(Args...),
+      const Extra&... extra) const {
+    m.def(name, f, extra...);
+    auto func = m.attr(name);
+    auto numba_support =
+        pybind11::module_::import("katana.native_interfacing.numba_support");
+    numba_support.attr("register_function")(
+        func, (uintptr_t)f, 0,
+        PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
+        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+  }
+
+public:
+  // Overload to handle non-capturing lambdas and static functions. These
+  // should all take self as the first argument.
+  template <typename Return, typename... Extra>
+  void operator()(
+      pybind11::module_& m, const char* name, Return (*pf)(Args...),
+      const Extra&... extra) const {
+    def_func(m, name, pf, extra...);
+  }
+
+  // Overload to handle functions values which are not yet pointers and forward
+  // them to the other overloads. This reduces the need for the user to
+  // explicitly write &.
+  template <typename Func, typename... Extra>
+  void operator()(
+      pybind11::module_& m, const char* name, Func&& f,
+      const Extra&... extra) const {
+    this->template operator()(m, name, &f, extra...);
+  }
+
+  // TODO(amp): The generated method wrappers used from numba code are created per
+  //  *signature*, not per function. So two functions with the same overall
+  //  signature (including containing class) will use the same wrapper. This
+  //  could possibly cause a megamorphic call site that defeats the CPU branch
+  //  predictor. The fix for this is to somehow distinguish different
+  //  functions/methods at the template level. This could be done with a counter
+  //  or name template argument of some kind.
+
+  // Overload to handle non-const methods on classes.
+  template <typename Cls, typename Return, typename ImplCls, typename... Extra>
+  void operator()(
+      pybind11::class_<Cls>& cls, const char* name,
+      Return (ImplCls::*pmf)(Args...), const Extra&... extra) const {
+    // This leaks a single pointer sized struct for each defined numba function.
+    // Repeated import could theoretically cause this to matter, but it's very
+    // unlikely.
+    def_class_method<Cls, Return>(
+        cls, name, pmf, new MemberFunction<ImplCls, Return, Args...>{pmf},
+        extra...);
+  }
+
+  // Overload to handle const methods on classes.
+  template <typename Cls, typename Return, typename ImplCls, typename... Extra>
+  void operator()(
+      pybind11::class_<Cls>& cls, const char* name,
+      Return (ImplCls::*pmf)(Args...) const, std::true_type,
+      const Extra&... extra) const {
+    // As above.
+    def_class_method<Cls, Return>(
+        cls, name, pmf, new ConstMemberFunction<ImplCls, Return, Args...>{pmf},
```

Comment:
How is `new ConstMemberFunc...` deleted?

Reply:
It is not. But there is a fix coming for this. Give me a couple of hours to setup the new design due to @danielmawhirter.